File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/bwtapa/domain/BWTapaObject.java
package com.zwstudio.logicpuzzlesandroid.puzzles.bwtapa.domain;

public abstract class BWTapaObject {
    public abstract String objTypeAsString();
    public static BWTapaObject objTypeFromString(String str) {
        switch (str) {
            case "marker":
                return new BWTapaMarkerObject();
            case "wall":
                return new BWTapaWallObject();
            default:
                return new BWTapaEmptyObject();
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/bwtapa/domain/BWTapaObject.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.bwtapa.domain

import com.zwstudio.logicpuzzlesandroid.common.domain.Position
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState

sealed class BWTapaObject {
    open fun objTypeAsString() = "empty"

    companion object {
        fun objTypeFromString(str: String) = when (str)  {
            "marker" -> BWTapaMarkerObject()
            "wall" -> BWTapaWallObject()
            else -> BWTapaEmptyObject()
        }
    }
}

class BWTapaEmptyObject : BWTapaObject()

class BWTapaHintObject(var state: HintState = HintState.Normal) : BWTapaObject() {
    override fun objTypeAsString() = "hint"
}

class BWTapaMarkerObject : BWTapaObject() {
    override fun objTypeAsString() = "marker"
}

class BWTapaWallObject(var state: HintState = HintState.Normal) : BWTapaObject() {
    override fun objTypeAsString() = "wall"
}

class BWTapaGameMove(val p: Position, var obj: BWTapaObject = BWTapaEmptyObject())
